### PR TITLE
Feature/cs 021 5/directory analyzer by path

### PR DESCRIPTION
**Pull request by luiggy-mamani-jalafund**
#### Resolved task: ([CS-021-5 [SP-1]: Implement directory analysis with DirectoryAnalyzerByPath](https://app.clickup.com/t/868aem2j9))
* * * *

## What's in this pull request?
Implemented directory reader/analyzer to build the AST based on the Java file paths.

## Type of change
- [X] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Unit tests have been added to test the behavior in a happy path and exception cases.

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [X] I have checked my code and corrected any misspellings

### CODE REVIEW

File: ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/application/exceptions/FileAnalyzerException.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package edu.usb.argos.ASTProcessor.application.exceptions;
+
+public class FileAnalyzerException extends Exception {
```
Comment (zamoranoalicia) [inline ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/application/exceptions/FileAnalyzerException.java]: our exceptions should always be Runtime exception

File: ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/infrastructure/analyzers/directoryAnalyzers/DirectoryAnalyzerByPath.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        List<Path> javaFilePaths = treeAnalyzer.getJavaFiles(path);
+        TAst ast;
+        List<TAst> astFiles = new ArrayList<>();
+
+        for (Path javaFilePath : javaFilePaths) {
+            try {
+                ast = fileAnalyzer.readFile(javaFilePath);
+                astFiles.add(ast);
+            } catch (FileAnalyzerException exception) {
+                log.error(exception.getMessage(), exception);
```
Comment (zamoranoalicia) [inline ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/infrastructure/analyzers/directoryAnalyzers/DirectoryAnalyzerByPath.java]: can we also throw the exception

**Review by zamoranoalicia — APPROVED** [review]
LGTM, good job, just some minor requested changes, please address them before merging

File: ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/application/analyzers/IDirectoryAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package edu.usb.argos.ASTProcessor.application.analyzers;
+
+import java.util.List;
+
+public interface IDirectoryAnalyzer<TSourceCode, TAst> {
```
Comment (axel-ayala-jalafund) [inline ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/application/analyzers/IDirectoryAnalyzer.java]: I think that our Trainer said, that don't use that, only use `T`, `F`, etc

File: ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/application/validators/IPathValidator.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package edu.usb.argos.ASTProcessor.application.validators;
+
+import java.nio.file.Path;
+
+public interface IPathValidator {
```
Comment (axel-ayala-jalafund) [inline ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/application/validators/IPathValidator.java]: For consistency with others interfaces, IPathValidator could be `IPathValidator<T>` to handle different types of paths

File: ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/infrastructure/utils/SourceTreeAnalyzer.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Slf4j
+@Component
+@AllArgsConstructor
+public class SourceTreeAnalyzer {
+
+    private final String JAVA_EXTENSION = ".java";
```
Comment (axel-ayala-jalafund) [inline ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/infrastructure/utils/SourceTreeAnalyzer.java]: That should be `static`

File: ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/application/analyzers/IDirectoryAnalyzer.java (hunk 3 above)
Comment (luiggy-mamani-jalafund) [reply ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/application/analyzers/IDirectoryAnalyzer.java]: done

File: ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/application/validators/IPathValidator.java (hunk 4 above)
Comment (luiggy-mamani-jalafund) [reply ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/application/validators/IPathValidator.java]: done

File: ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/infrastructure/utils/SourceTreeAnalyzer.java (hunk 5 above)
Comment (luiggy-mamani-jalafund) [reply ASTProcessor/src/main/java/edu/usb/argos/ASTProcessor/infrastructure/utils/SourceTreeAnalyzer.java]: done

**Review by diego-centella-jalafund — APPROVED** [review]

**Review by axel-ayala-jalafund — APPROVED** [review]

**Pull request merged**